src/server/scripts/Spells/spell_generic.cpp

```diff
@@ -1614,6 +1614,119 @@ class spell_gen_elune_candle : public SpellScript
     }
 };
 
+// 50051 - Ethereal Pet Aura
+enum EtherealPet
+{
+    NPC_ETHEREAL_SOUL_TRADER        = 27914,
+
+    SAY_STEAL_ESSENCE               = 1,
+    SAY_CREATE_TOKEN                = 2,
+
+    SPELL_PROC_TRIGGER_ON_KILL_AURA = 50051,
+    SPELL_ETHEREAL_PET_AURA         = 50055,
+    SPELL_STEAL_ESSENCE_VISUAL      = 50101,
+    SPELL_CREATE_TOKEN              = 50063
+};
+
+// 50051 - Ethereal Pet Aura
+class spell_ethereal_pet_aura : public AuraScript
+{
+    PrepareAuraScript(spell_ethereal_pet_aura);
+
+    bool CheckProc(ProcEventInfo& eventInfo)
+    {
+        uint32 levelDiff = std::abs(GetTarget()->getLevel() - eventInfo.GetProcTarget()->getLevel());
+            return levelDiff <= 9;
+    }
+
+    void HandleProc(AuraEffect const* /*aurEff*/, ProcEventInfo& eventInfo)
+    {
+        PreventDefaultAction();
+
+        std::list<Creature*> MinionList;
+        GetUnitOwner()->GetAllMinionsByEntry(MinionList, NPC_ETHEREAL_SOUL_TRADER);
+
+        Creature* soulTrader = MinionList.front()->ToCreature();
+        soulTrader->AI()->Talk(SAY_STEAL_ESSENCE);
+        soulTrader->CastSpell(eventInfo.GetProcTarget(), SPELL_STEAL_ESSENCE_VISUAL);
+    }
+
+    void Register() override
+    {
+        DoCheckProc += AuraCheckProcFn(spell_ethereal_pet_aura::CheckProc);
+        OnEffectProc += AuraEffectProcFn(spell_ethereal_pet_aura::HandleProc, EFFECT_0, SPELL_AURA_PROC_TRIGGER_SPELL);
+    }
+};
+
+// 50052 - Ethereal Pet onSummon
+class spell_ethereal_pet_onsummon : public SpellScript
+{
+    PrepareSpellScript(spell_ethereal_pet_onsummon);
+
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_PROC_TRIGGER_ON_KILL_AURA });
+    }
+
+    void HandleScriptEffect(SpellEffIndex /*effIndex*/)
+    {
+        Unit* target = GetHitUnit();
```

Comment:
Not needed.. do GetHitUnit()->AddAura

Reply:
AddAura requires target.
```cpp
Aura* AddAura(uint32 spellId, Unit* target);
Aura* AddAura(SpellInfo const* spellInfo, uint8 effMask, Unit* target);
```

Reply:
Ah crap, didn't see the second parameter, sry

Reply:
Yeah well then let it stay target or change both to GetHitUnit(), either one or the other. Sry for the comment at all :D

Reply:
No problem :+1:

Reply:
wait, add aura? please no !
Use cast spell instead.

```cpp

void HandleScriptEffect(SpellEffIndex /*effIndex*/)
{
    GetHitUnit()->CastSpell(GetHitUnit(), SPELL_ETHEREAL_PET_AURA, true);
}
```

Reply:
Fixed. Thanks!

Reply:
remove ``Unit* target = GetHitUnit();`` then

Reply:
`GetHitUnit()` is used twice. Isn't it better to move it into a variable?

Reply:
OK :)